src/ripple/overlay/README.md

```diff
@@ -147,41 +134,128 @@ Content-Type: application/json
+    for node public keys.
 
 * `Session-Signature`
 
     This field must be present. It contains a cryptographic token formed
     from the SHA512 hash of the shared data exchanged during SSL handshaking.
-    For more details see the corresponding source code.
+    For more details see "Session Signature" below.
+
+    The hash is encoded using **Base64** encoding, but a future release of the
+    software will switch to supporting hexadecimal encoding.
 
 * `Crawl` (optional)
 
     If present, and the value is "public" then neighbors will report the IP
     address to crawler requests. If absent, neighbor's default behavior is to
     not report IP addresses.
 
+* `Closed-Ledger` (optional)
+
+    If present, identifies the hash of the last ledger that the sending server
+    considers to be closed.
+
+    The hash is encoded using **Base64** encoding, but a future release of the
+    software will switch to supporting hexadecimal encoding.
+
+* `Previous-Ledger` (optional)
+
+    If present, identifies the hash of the parent ledger that then sending server
+    considers to be closed.
+
+    The hash is encoded using **Base64** encoding, but a future release of the
+    software will switch to supporting hexadecimal encoding.
+
 * _User Defined_ (Unimplemented)
 
-    The rippled operator may specify additional, optional fields and values
-    through the configuration. These headers will be transmitted in the
-    corresponding request or response messages.
+    An implementation or operator may specify additional, optional fields
+    and values. These headers will be transmitted in the corresponding request
+    or response messages. They will be ignored by implementations that do not
+    understand the header.
+
+### Session Signature
+
+Even for SSL/TLS encrypted connections, it is possible for an attacker to mount
+relatively inexpensive MITM attacks that can be extremely hard to detect and
+may afford the attacker the ability to intelligently tamper with messages
+exchanged between the two endpoints.
+
+This risk can be mitigated if at least one side has a certificate from a certificate
+authority trusted by the other endpoint, but having a certificate is not always
+possible (or even desirable) in a decentralized and permissionless network.
+
+Ultimately, the goal is to ensure that two endpoints A and B know that they are
+talking directly to each other over a single end-to-end SSL/TLS session instead
+of two separate SSL/TLS sessions, with an attacker acting as a proxy.
+
+The XRP Ledger protocol prevents this attack by leveraging the fact that the two
+servers each have a node identity, in the form of **`secp256k1`** keypairs, and
+use that to strongly bind the SSL/TLS session to the node identities of each of
+the two servers at the end of the SSL/TLS session.
+
+To do this we "reach into" the SSL/TLS session, and extract the **`finished`**
+messages for the local and remote endpoints, and combine them to generate a unique
```

Comment:
Urgh, I tried to implement this in Python with `pyOpenSSL` and `cryptography` (and `base58`, since the node public key is encoded that way - it would be nice to have _everything_ in either hex, base64 or base58...).

`rippled` seems to just drop connection attempts it doesn't like/understand. It would be nice to have at least a HTTP 400 error or similar returned instead.

Reply:
Try testing against this branch. You should _reliably_ get an error message that is descriptive. If you don't, let me know.